# k-Padre

This repository contains a suite of tools for analyzing allele-specific genomic data from hybrid samples. It includes tools for allelic read assignment on genomic, gene expression, and DNA-methylation data from bisulphite sequencing. It also includes a utility to model the effects of tissue contamination in quimeric tissues like placentas. These bioinformatic pipeline, is optimized for genomic analyisis using a diploid genome assembly, but provides alternatives to use a haploid genome and a set of diagnostic SNPs.

## Table of Contents

- [Features](#features)
- [Dependencies](#dependencies)
- [Installation](#installation)
  
- [Usage](#usage)
- [Input/Output Format](#inputoutput-format)
- [Examples](#examples)
- [Contributing](#contributing)
- [License](#license)
- [Contact](#contact)

## Features

- Classifies genoic sequencing reads by allele 
- Supports DNA-methylation data generated by sulphite sequencing.
- Models the effect of tissue contamination on quimeric tissues.
- Compatible with HISAT2, and WGBS alignment tools

## Dependencies

This script requires [`samtools`](http://www.htslib.org/) to be installed. Samtools is included in the kPadre-env.yalm file for conda, but a custom installation can be used as long as samtools is in your system PATH.

## Installation

Clone the repository:

```bash
git clone https://github.com/LF_Rodriguez/kPadre.git
cd k-Padre

# Optional conda environment for samtools
conda env create -f kPadre-env.yml
conda activate kPadre-env
```
## Usage

## Inputoutput-format

## examples

## Contributing

## License

## contact

